File: grpc/nodejs/src/exceptions.js
/**
 * Custom exception types for the freeplane-grpc-client library.
 *
 * Exception hierarchy:
 *   FreeplaneGrpcError (base)
 *     ├── FreeplaneConnectionError  (network/gRPC transport failures)
 *     └── FreeplaneOperationError   (server-reported operation failures)
 *           ├── NodeNotFoundError   (requested node does not exist)
 *           └── MindMapError        (map-level operation failures)
 */

/**
 * Base exception for all Freeplane gRPC errors.
 * @extends Error
 */
class FreeplaneGrpcError extends Error {
  constructor(message) {
    super(message);
    this.name = 'FreeplaneGrpcError';
  }
}

/**
 * Raised when a connection to the Freeplane gRPC server fails.
 * @extends FreeplaneGrpcError
 */
class FreeplaneConnectionError extends FreeplaneGrpcError {
  constructor(message) {
    super(message);
    this.name = 'FreeplaneConnectionError';
  }
}

/**
 * Raised when a gRPC operation fails (server reported failure).
 * @extends FreeplaneGrpcError
 */
class FreeplaneOperationError extends FreeplaneGrpcError {
  constructor(message) {
    super(message);
    this.name = 'FreeplaneOperationError';
  }
}

/**
 * Raised when a requested node is not found.
 * @extends FreeplaneOperationError
 */
class NodeNotFoundError extends FreeplaneOperationError {
  constructor(message) {
    super(message);
    this.name = 'NodeNotFoundError';
  }
}

/**
 * Raised when a map-level operation fails.
 * @extends FreeplaneOperationError
 */
class MindMapError extends FreeplaneOperationError {
  constructor(message) {
    super(message);
    this.name = 'MindMapError';
  }
}

module.exports = {
  FreeplaneGrpcError,
  FreeplaneConnectionError,
  FreeplaneOperationError,
  NodeNotFoundError,
  MindMapError,
};

File: grpc/kotlin/src/main/kotlin/org/freeplane/grpc/Exceptions.kt
package org.freeplane.grpc

/**
 * Base exception for all Freeplane gRPC errors.
 */
open class FreeplaneGrpcError(message: String, cause: Throwable? = null) : RuntimeException(message, cause) {
    override val message: String get() = super.message ?: ""
}

/**
 * Indicates a failure to connect or communicate with the server.
 * Covers gRPC transport failures (UNAVAILABLE, DEADLINE_EXCEEDED, RESOURCE_EXHAUSTED).
 */
class FreeplaneConnectionError(message: String, cause: Throwable? = null) : 
    FreeplaneGrpcError(message, cause)

/**
 * Indicates a failure reported by the server during an operation.
 * Covers server-reported failures (success=false).
 */
open class FreeplaneOperationError(message: String, cause: Throwable? = null) : 
    FreeplaneGrpcError(message, cause)

/**
 * Indicates that a requested node was not found.
 */
class NodeNotFoundError(message: String, cause: Throwable? = null) : 
    FreeplaneOperationError(message, cause)

/**
 * Indicates a failure at the mind map level.
 */
class MindMapError(message: String, cause: Throwable? = null) : 
    FreeplaneOperationError(message, cause)
